crates/gpui/src/assets.rs

```diff
@@ -35,40 +43,135 @@ impl AssetSource for () {
 pub struct ImageId(usize);
 
 /// A cached and processed image.
-pub struct ImageData {
-    /// The ID associated with this image
-    pub id: ImageId,
-    data: ImageBuffer<Bgra<u8>, Vec<u8>>,
+pub enum ImageData {
+    /// A raster image.
+    Raster {
+        /// The unique identifier for this image.
+        id: ImageId,
+        /// The image buffer for this image.
+        data: ImageBuffer<Bgra<u8>, Vec<u8>>,
+    },
+    /// A vector image.
+    Vector {
+        /// The unique identifier for this image.
+        id: ImageId,
+        /// The SVG tree for this image.
+        data: resvg::usvg::Tree,
+        /// Rendered image cache.
+        rendered: Arc<Mutex<HashMap<Size<DevicePixels>, FetchImageTask>>>,
+    },
 }
 
+static NEXT_ID: AtomicUsize = AtomicUsize::new(0);
 impl ImageData {
     /// Create a new image from the given data.
     pub fn new(data: ImageBuffer<Bgra<u8>, Vec<u8>>) -> Self {
-        static NEXT_ID: AtomicUsize = AtomicUsize::new(0);
-
-        Self {
+        Self::Raster {
             id: ImageId(NEXT_ID.fetch_add(1, SeqCst)),
             data,
         }
     }
 
+    /// Try to create a new image from the given bytes.
+    pub fn try_from_bytes(bytes: &[u8]) -> Result<Self, ImageCacheError> {
+        if let Ok(format) = image::guess_format(bytes) {
+            let data = image::load_from_memory_with_format(bytes, format)?.into_bgra8();
+            Ok(Self::Raster {
+                id: ImageId(NEXT_ID.fetch_add(1, SeqCst)),
+                data,
+            })
+        } else {
+            let data = resvg::usvg::Tree::from_data(
+                bytes,
+                &resvg::usvg::Options::default(),
+                &resvg::usvg::fontdb::Database::default(),
```

Comment:
This will create an empty font database. I am not sure what exactly usvg uses it for tho - it may be fine. Alternatively we can create a database that contains the system fonts https://docs.rs/fontdb/latest/fontdb/struct.Database.html#method.load_system_fonts and pass that in here.

Reply:
Ok yeah that definitely makes sense. Otherwise SVGs with text embedded just won’t display the text. Do you think we need to cache the fontdb? No idea how expensive it is to initialize it for each SVGs.

Reply:
Depends on how many fonts the user has on their system. I think caching would be better, since it needs to read quite a few files every time it loads the fonts.

Reply:
@niklaswimmer do you think a OnceLock would be acceptable for caching the fontdb? I am still quite new to Rust so just want to make sure we are not endlessly iterating on some minor detail.